crates/metadata-store/src/local/grpc/net_util.rs

```diff
@@ -0,0 +1,238 @@
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use http::Uri;
+use hyper::body::{Body, Incoming};
+use hyper::rt::{Read, Write};
+use hyper_util::rt::TokioIo;
+use restate_core::{cancellation_watcher, task_center, ShutdownError, TaskCenter, TaskKind};
+use restate_types::errors::GenericError;
+use restate_types::net::{AdvertisedAddress, BindAddress};
+use std::fmt::Debug;
+use std::future::Future;
+use std::net::SocketAddr;
+use std::path::PathBuf;
+use std::time::Duration;
+use tokio::io;
+use tokio::net::{TcpListener, UnixListener, UnixStream};
+use tokio_util::net::Listener;
+use tonic::transport::{Channel, Endpoint};
+use tracing::{debug, info, instrument, Span};
+
+pub fn create_tonic_channel_from_advertised_address(
+    address: AdvertisedAddress,
+) -> Result<Channel, http::Error> {
+    let channel = match address {
+        AdvertisedAddress::Uds(uds_path) => {
+            // dummy endpoint required to specify an uds connector, it is not used anywhere
+            Endpoint::try_from("http://127.0.0.1")
+                .expect("/ should be a valid Uri")
+                .connect_with_connector_lazy(tower::service_fn(move |_: Uri| {
+                    let uds_path = uds_path.clone();
+                    async move {
+                        Ok::<_, io::Error>(TokioIo::new(UnixStream::connect(uds_path).await?))
+                    }
+                }))
+        }
+        AdvertisedAddress::Http(uri) => {
+            // todo: Make the channel settings configurable
+            Channel::builder(uri)
+                .connect_timeout(Duration::from_secs(5))
+                // todo: configure the channel from configuration file
+                .http2_adaptive_window(true)
+                .connect_lazy()
+        }
+    };
+    Ok(channel)
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("failed binding to address '{address}': {source}")]
+    TcpBinding {
+        address: SocketAddr,
+        #[source]
+        source: io::Error,
+    },
+    #[error("failed opening uds '{uds_path}': {source}")]
+    UdsBinding {
+        uds_path: PathBuf,
+        #[source]
+        source: io::Error,
+    },
+    #[error("failed handling hyper connection: {0}")]
+    HandlingConnection(#[from] GenericError),
+    #[error("failed listening on incoming connections: {0}")]
+    Listening(#[from] io::Error),
+    #[error(transparent)]
+    Shutdown(#[from] ShutdownError),
+}
+
+#[instrument(level = "info", skip_all, fields(server_name = %server_name, uds.path = tracing::field::Empty, net.host.addr = tracing::field::Empty, net.host.port = tracing::field::Empty))]
+pub async fn run_hyper_server<S, B>(
```

Comment:
Now that I look at this, I think the admin server should use this same code to serve a server.

Perhaps how about you extract this in a separate small crate, like `restate-hyper-util`, where you have this `serve` function that serves on the task center?